src/OpenFeature/EventExecutor.cs

```diff
@@ -0,0 +1,368 @@
+                }
+
+                this._namedProviderReferences[client] = newProvider;
+
+                this.StartListeningAndShutdownOld(newProvider, oldProvider);
+            }
+        }
+
+        private void StartListeningAndShutdownOld(FeatureProviderReference newProvider, FeatureProviderReference oldProvider)
+        {
+            // check if the provider is already active - if not, we need to start listening for its emitted events
+            if (!this.IsProviderActive(newProvider))
+            {
+                this._activeSubscriptions.Add(newProvider);
+                var featureProviderEventProcessing = new Thread(this.ProcessFeatureProviderEventsAsync);
+                featureProviderEventProcessing.Start(newProvider);
+            }
+
+            if (oldProvider != null && !this.IsProviderBound(oldProvider))
+            {
+                this._activeSubscriptions.Remove(oldProvider);
+                var channel = oldProvider.Provider.GetEventChannel();
+                if (channel != null)
+                {
+                    channel.Writer.WriteAsync(new ShutdownSignal());
+                }
+            }
+        }
+
+        private bool IsProviderBound(FeatureProviderReference provider)
+        {
+            if (this._defaultProvider == provider)
+            {
+                return true;
+            }
+            foreach (var providerReference in this._namedProviderReferences.Values)
+            {
+                if (providerReference == provider)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private bool IsProviderActive(FeatureProviderReference providerRef)
+        {
+            return this._activeSubscriptions.Contains(providerRef);
+        }
+
+        private void EmitOnRegistration(FeatureProviderReference provider, ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            if (provider == null)
+            {
+                return;
+            }
+            var status = provider.Provider.GetStatus();
+
+            var message = "";
+            if (status == ProviderStatus.Ready && eventType == ProviderEventTypes.ProviderReady)
+            {
+                message = "Provider is ready";
+            }
+            else if (status == ProviderStatus.Error && eventType == ProviderEventTypes.ProviderError)
+            {
+                message = "Provider is in error state";
+            }
+            else if (status == ProviderStatus.Stale && eventType == ProviderEventTypes.ProviderStale)
+            {
+                message = "Provider is in stale state";
+            }
+
+            if (message != "")
+            {
+                try
+                {
+                    handler.Invoke(new ProviderEventPayload
+                    {
+                        ProviderName = provider.Provider.GetMetadata().Name,
```

Comment:
```suggestion
                        ProviderName = provider.Provider?.GetMetadata()?.Name,
```

Technically this shouldn't be an issue, but I think we should be really defensive in the SDK.